lib/game/client/engine/phaser/phaser-scene.driver.js

```diff
@@ -0,0 +1,191 @@
+const {BaseSceneDriver} = require("../base-scene-driver");
+const {Scene, Geom} = require('phaser');
+const {sc} = require("@reldens/utils");
+
+class PhaserSceneDriver extends BaseSceneDriver
+{
+    constructor(config)
+    {
+        super();
+        this.driverName = "Phaser Scene Driver";
+        this.scene = new WrapperScene(config);
+    }
+
+    setUpPreloadCallback(callback)
+    {
+        this.getScene().setUpPreloadCallback(callback);
+    }
+
+    setUpCreateCallback(callback)
+    {
+        this.getScene().setUpCreateCallback(callback);
+    }
+
+    loadHTML(key, url, xhrSettings)
+    {
+        this.getScene().load.html(key, url, xhrSettings);
+    }
+
+    loadTilemapTiledJSON(key, url, xhrSettings)
+    {
+        this.getScene().load.tilemapTiledJSON(key, url, xhrSettings);
+    }
+
+    loadSpritesheet(key, url, frameConfig, xhrSettings)
+    {
+        this.getScene().load.spritesheet(key, url, frameConfig, xhrSettings);
+    }
+
+    loadImage(key, url, xhrSettings)
+    {
+        this.getScene().load.image(key, url, xhrSettings);
+    }
+
+    loadOn(event, fn, context)
+    {
+        this.getScene().load.on(event, fn, context);
+    }
+
+    getCameraManager()
+    {
+        return this.getScene().cameras;
+    }
+
+    getMainCamera()
+    {
+        return this.getCameraManager().main;
+    }
+
+    getMainCameraWidth()
+    {
+        return this.getMainCamera().width;
+    }
+
+    getMainCameraHeight()
+    {
+        return this.getMainCamera().height;
+    }
+
+    addGraphics()
+    {
+        return this.getScene().add.graphics();
+    }
+
+    addDomCreateFromCache(x, y, config)
+    {
+        return this.getScene().add.dom(x, y, config.element, config.style, config.innerText).createFromCache(config.the);
+    }
+
+    getCacheHtml(key)
+    {
+        return this.getScene().cache.html.get(key);
+    }
+
+    createAnimation(config)
+    {
+        return this.getScene().anims.create(config);
+    }
+
+    generateAnimationFrameNumbers(key, config)
+    {
+        return this.getScene().anims.generateFrameNumbers(key, config);
+    }
+
+    addText(x, y, text, style)
+    {
+        return this.getScene().add.text(x, y, text, style);
+    }
+
+    getRectangle()
+    {
+        return Geom.Rectangle;
+    }
+
+    getScenePlugin()
+    {
+        return this.getScene().scene;
+    }
+
+    getChildren()
+    {
+        return this.getScene().children;
+    }
+
+    inputKeyboardOn(event, key)
+    {
+        return this.getScene().input.keyboard.on(event, key);
+    }
+
+    setUiElement(uiName, value)
+    {
+        this.getScene().setUiElement(uiName, value);
+    }
+
+    getUiElement(uiName, logError = true)
+    {
+        return this.getScene().getUiElement(uiName, logError);
+    }
+
+    getAllUiElements()
+    {
+        return this.getScene().getAllUiElements();
+    }
+
+    getScene()
+    {
+        return this.scene;
+    }
+}
+
+class WrapperScene extends Scene
```

Comment:
- we shouldn't extend Phaser Scene, all the methods should be native in our SceneDriver
- our SceneDriver should create the Phaser.Scene on it's own constructor and save it as property, you can call it "scene", and you can pass the contructor "props" entirely to the Phase.Scene.
- that way we will be able to call our SceneDriver.preload() which will internally call this.scene.preload(), this way you can see how nothing of Phase.Scene will be out of this driver (the storage drivers on the admin works this way, you can take a look how the driver instance is created and how the resources are assigned.
- note, we won't necessarily have ALL the Phaser.Scene methods in our SceneDriver, we should only keep the "main" ones (by this I mean the ones that runs specific functions).
some examples:
-- preload() and create() are very basic methods that we will use not matter the engine we put behind because in our SceneDriver logic we need to allow other devs to specify stuff to happe on preload or create
-- or let's say we have a createUiElement() method, that should be available in our SceneDriver, but we don't care how later the specific driver use the engine for that action.

Reply:
"that way we will be able to call our SceneDriver.preload() which will internally call this.scene.preload(), this way you can see how nothing of Phase.Scene will be out of this driver (the storage drivers on the admin works this way, you can take a look how the driver instance is created and how the resources are assigned." as far as I know, scene.preload is not called by us anywhere in the code. It is called automatically inside de Phaser engine when you add/load de scene to the engine. _(Phaser Doc ["Can be defined on your own Scenes. Use it to load assets. This method is called by the Scene Manager, after init() and before create(), only if the Scene has a LoaderPlugin. After this method completes, if the LoaderPlugin's queue isn't empty, the LoaderPlugin will start automatically."](https://newdocs.phaser.io/docs/3.55.2/Phaser.Types.Scenes.ScenePreloadCallback))_ 

Because it is a callback ¿the only way? to run our code inside the preload and create methods we have to extend the Phaser.Scene class, override the methods and pass our callback, so the Phase.SceneManager runs them.

Reply:
hmm... good point, I've totally forgot about that..... 
BUT, a better approach that I think it should work, and I will welcome you, to the world of bind: 
https://www.w3schools.com/js/js_function_bind.asp

- We can create a Phaser.Scene instance in our PhaserDriver (like in the engine), you don't need to put ANYTHING on it, just a plain new Scene().
- Then we can create our own PhaserSceneDriver, which will contain any required methods we will need.
- For last, we bind the Scene.preload() to our classe, like: 
```
let phaserScene = new Scene();
let ourDriver = new ScenePreloadDriver();
phaserScene.preload.bind(ourDriver);
```

Give it a try, if that doesn't work let's schedule a call :) and if it works, let's schedule a call anyway so we can catch up :D

Reply:
Is possible the commented approach doesn't work, if that's the case then keep the wrapper for now but let's move it to it's own file

Reply:
OK! Gonna investigate abount bind and other magical functionality of JS, I love them (irony). xD